### PR TITLE
adding cmoka test framework

### PR DESCRIPTION
**Pull request by vhvb1989**
This is a proposal of using cmocka for unit testing https://cmocka.org/
This would allow us to mock functions and to manage multiple tests binaries easily
It also offer a nice api for assertions (numbers, bools, pointers, etc). 
And it can also be used to expect fails.

I think the only cons would be that we would need to install cmocka on CI for unit testing. (adding the script to download, compile and install)

let me know what you think
I am adding just a few tests as example.. we would need to migrate all existing unit tests.. or at least start adding any new test with this and leave previous test binaries as it is

### CODE REVIEW

**Comment by danieljurek** [comment]
Now includes changes to the pipeline yaml to perform: 
* Build validation -- install only curl and test build (i.e. `UNIT_TESTING: OFF`)
* Build/Test validation -- install curl and test dependencies, build (with `UNIT_TESTING: ON`) , run ctest

**Comment by vhvb1989** [comment]
> Before we commit this we need to ensure its an acceptable open source project to take a dependency on

it is acceptable.  Here is some popular projects using cmocka:

image

**Comment by joshfree** [comment]
@vhvb1989 This seems like a good solution overall.  Lets chat in our standup real quick.  In general for any third party dependency we need to double check that this is acceptable for the IoT team, meets our licensing requirements, and I’d like to confirm that we’ve done some due diligence on the various popular unit test frameworks for C and we believe this is the best solution for us (ex:  http://www.throwtheswitch.org/comparison-of-unit-test-frameworks)

**Comment by vhvb1989** [comment]
> @vhvb1989 This seems like a good solution overall. Lets chat in our standup real quick. In general for any third party dependency we need to double check that this is acceptable for the IoT team, meets our licensing requirements, and I’d like to confirm that we’ve done some due diligence on the various popular unit test frameworks for C and we believe this is the best solution for us (ex: http://www.throwtheswitch.org/comparison-of-unit-test-frameworks)


@joshfree I put together a comparison of cMocka against other popular C/C++ test frameworks here:
https://microsoft.sharepoint.com/:o:/t/AzureDeveloperExperience/EueKRWtSWF5OoK8MuTl5zu8BnHag7r9igrKEZQZJXxp1EA?e=QNLG8R

File: sdk/core/core/test/mocka_json.c [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  assert_int_equal(az_span_capacity(token.value.string), 0);
+  assert_int_equal(az_span_length(token.value.string), 0);
+
+  assert_ptr_equal(az_span_ptr(token.value.span), NULL);
+  assert_int_equal(az_span_capacity(token.value.span), 0);
+  assert_int_equal(az_span_length(token.value.span), 0);
+}
+
+static void test_json_token_boolean(void ** state) {
+  (void)state; /* unused */
```
Comment (antkmsft) [inline sdk/core/core/test/mocka_json.c]: `/* unused */` comment is redundant, in my opinion. If you know the language, there's really no other explanation to the `(void)state; line other than it is to clearly indicate that it is not used, but suppress warning.

File: sdk/core/core/test/mocka_json.c [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+int main(void) {
+  const struct CMUnitTest tests[] = {
+    cmocka_unit_test(test_json_token_null),
+    cmocka_unit_test(test_json_parser_init),
+    cmocka_unit_test(test_json_token_boolean),
+    cmocka_unit_test(test_json_token_number),
+  };
+
+  return cmocka_run_group_tests(tests, NULL, NULL);
```
Comment (antkmsft) [inline sdk/core/core/test/mocka_json.c]: Are you going to port over all other test cases we have to use cmocka as part of this PR?

**Comment by antkmsft** [comment]
Also please don't forget to run code formatting on all the .c/.h files that you're modifying

File: sdk/core/core/test/mocka_json.c (hunk 2 above)
Comment (vhvb1989) [reply sdk/core/core/test/mocka_json.c]: that might take a while.
We could take this a the way to add any new test and leave the previous one like it is
Both tests can work together with no problem

**Comment by gilbertw** [comment]
@vhvb1989 lets put the cmoka tests under test folders.
Thus we would have
```
sdk/core/core/test/cmoka/....
sdk/storage/blobs/test/cmoka/...
...
```

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml [hunk 3]
```diff
@@ -106,3 +164,12 @@ jobs:
...
     displayName: ctest
+  # Publish Test Results
+  # Publish test results to Azure Pipelines
+  - task: PublishTestResults@2
+    inputs:
+      testResultsFormat: 'JUnit'
+      testResultsFiles: '**/*-test-results.xml' 
+      mergeTestResults: true
+      testRunTitle: 'Azure SDK C99 - $(vm.image)_$(VCPKG_DEFAULT_TRIPLET)'
+    condition: and(succeeded(), contains(variables['vcpkg.deps'], 'cmocka')))
```
Comment (danieljurek) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml]: We always want to upload test results if they're available so this is one of the few exceptions where we shouldn't be checking `succeeded()`

Comment (vhvb1989) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml]: I was trying to avoid failing here if no test results files were generated at all

**Pull request merged**